currency_rate_update/currency_rate_update.py

```diff
@@ -731,70 +710,127 @@ def get_updated_currency(self, currency_array, main_currency,
-                    reported by Bank of Canada.' % str(curr))
-
-            _logger.debug("BOC sent a valid RSS file for: " + curr)
-
-            # check for valid exchange data
-            if (dom.entries[0].cb_basecurrency == main_currency) and \
-                    (dom.entries[0].cb_targetcurrency == curr):
-                rate = dom.entries[0].cb_exchangerate.split('\n', 1)[0]
-                rate_date_datetime = parser.parse(dom.entries[0].updated)\
-                    .astimezone(pytz.utc).replace(tzinfo=None)
-                self.check_rate_date(rate_date_datetime, max_delta_days)
-                self.updated_currency[curr] = rate
-                _logger.debug("BOC Rate retrieved : %s = %s %s" %
-                              (main_currency, rate, curr))
-            else:
-                _logger.error(
-                    "Exchange data format error for Bank of Canada -"
-                    "%s. Please check provider data format "
-                    "and/or source code." % curr)
-                raise osv.except_osv('Error !',
-                                     'Exchange data format error for\
-                                     Bank of Canada - %s !' % str(curr))
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    @property
+    def base_currency(self):
+        return self.dom.entries[0].cb_basecurrency
+
+    @property
+    def target_currency(self):
+        return self.dom.entries[0].cb_targetcurrency
+
+    @property
+    def exchange_rate(self):
+        return self.dom.entries[0].cb_exchangerate.split("\n", 1)[0]
+
+    @property
+    def updated(self):
+        return self.dom.entries[0].updated
+
+    @property
+    def date_time(self):
+        parsed_datetime = dateutil.parser.parse(self.updated)
+        return parsed_datetime.astimezone(pytz.utc).replace(tzinfo=None)
+
+
+class BankOfCanadaGetter(CurrencyGetterInterface):
+    """Implementation of Currency_getter_factory interface
+    for Bank of Canada RSS service
+
+    """
+
+    def get_updated_currency(self, currencies, main_currency, max_delta_days):
+        """Implementation of abstract method of Currency_getter_interface
+
+        :param currencies: List of currency name
+        :type currencies: list of unicode
+        :param main_currency: Name of Company's currency
+        :type main_currency: unicode
+        :param max_delta_days: Maximum allowable days of unsync
+        :type max_delta_days: int
+        :return: Updated list of currencies and their rates and log
+        :rtype: dict, str
+        :raises: orm.except_orm when there is a format error
+        """
+
+        errors = []
+
+        for currency in currencies:
+            # We do not want to update the main currency
+            if currency == main_currency:
+                continue
+            try:
+                self.validate_currency(currency)
+                with BankOfCanadaConnection(currency) as conn:
```

Comment:
I chose that direction in case the implementation were to require cleanup in future fixes.